src/AutoRest.CSharp/LowLevel/Output/ExpressionTypeProvider.cs

```diff
@@ -0,0 +1,53 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using AutoRest.CSharp.Common.Output.Models;
+using AutoRest.CSharp.Generation.Types;
+using AutoRest.CSharp.Input.Source;
+using AutoRest.CSharp.Output.Models.Types;
+
+namespace AutoRest.CSharp.LowLevel.Output
+{
+    // TODO -- eventually we should be able to move this class to a more common place, or just combine everything in this class into TypeProvider
+    internal abstract class ExpressionTypeProvider : TypeProvider
+    {
+        protected ExpressionTypeProvider(string defaultNamespace, SourceInputModel? sourceInputModel) : base(defaultNamespace, sourceInputModel)
+        {
+        }
+
+        private IEnumerable<string>? _usings;
+        public IEnumerable<string> Usings => _usings ??= BuildUsings();
+
+        protected virtual IEnumerable<string> BuildUsings()
+        {
+            yield break;
+        }
+
+        public virtual CSharpType? Inherits => null;
+
+        private IEnumerable<CSharpType>? _implements;
+        public virtual IEnumerable<CSharpType> Implements => _implements ??= BuildImplements();
```

Comment:
Do we need any of these methods to be lazy in any of the derived types?

Reply:
I think some of them could and some of them have to be lazy loaded.
For instance, `Inherits` could almost be determined when we construct this type provider, as well as Implements.
But others like the members of this type, could not be determined when we constructed this type, just like how we did the `BuildProperties` in models.
Let us take the `Resource` as an example, the methods of it is very likely to involve other resources or collections that is being constructed.
I am not 100% sure until we get there to make other types like this, therefore I could change them now since they really could be fully determined in the ctor here for the test case type provider. But if some cross reference issues occur in the future (most likely to happen in mgmt plane code), we might need to make them lazy loaded.

Also Michael is kind of a fan of lazy loading them - we even have an issue for that to convert dpg things to lazy loaded here:  https://github.com/Azure/autorest.csharp/issues/3371
This issue was created quite a time ago that he found in some circumstances when debugging things, the variables in the ctor is not really fully initialized.

Reply:
Also any suggestions for the name of this type? I only leave this name as a temporary placeholder

Reply:
Updated